include/envoy/config/incremental_subscription.h

```diff
@@ -0,0 +1,95 @@
+#pragma once
+
+#include <string>
+#include <vector>
+
+#include "envoy/common/exception.h"
+#include "envoy/common/pure.h"
+#include "envoy/stats/stats_macros.h"
+
+#include "common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Config {
+
+template <class ResourceType> class IncrementalSubscriptionCallbacks {
+public:
+  typedef Protobuf::RepeatedPtrField<ResourceType> ResourceVector;
+
+  virtual ~IncrementalSubscriptionCallbacks() {}
+
+  /**
+   * Called when an incremental configuration update is received.
+   * @param added_resources resources newly added since the previous fetch.
+   * @param removed_resources names of resources that this fetch instructed to be removed.
+   * @param version_info update version.
+   * @throw EnvoyException with reason if the incremental config changes are rejected. Otherwise the
+   *        changes are accepted. Accepted changes have their version_info reflected in subsequent
```

Comment:
One of the asks we've had is that we support partial updates, where we ignore (and don't ack specifically) those resources that fail but allow other to precede. Do you plan on supporting this?

Reply:
Oh interesting. Hmm, maybe a new exception type like XdsUpdateException that would list the failed resources? onConfigUpdate would proceed as if just the successfully updated resources were present in the update, and onConfigUpdateFailed would get the list of failed ones to further deal with? Does that sound vaguely on the right track?

(I don't have a sense of what a failure is likely to be, and in particular, whether it makes sense to trigger retries in response to partial failures, or just to sort of log the error and move on).

Reply:
Yeah, I think that when we are updating individual resources, e.g. a single cluster, it's fine to still throw an `EnvoyException`. We could then aggregate them in the loop that is doing these updates into an `XdsUpdateException` and both use this to to NACK those resources on the wire and for `onConfigUpdateFailed`.

An example of a partial failure may be during an RDS update, where 10 `RouteConfiguration`s are updated, 9 of them are fine, but 1 of them references a non-existent `Cluster`.

Reply:
I looked into this more closely. I see what you mean about aggregating individual normal exceptions; I see where in the code it would all go, and it seems quite simple indeed. I think the protocol is going to need to be more expressive. Currently, you just set the error_detail field (google.rpc.Status) to indicate that the message is a NACK. Either that should be a map from resources to errors, or that map should be an additional field. The former is cleaner unless there are cases where it's meaningful to distinguish "the whole update had a problem" from "every updated resource had a problem".

Reply:
A repeated status field sounds good. Can you open an issue to discuss whether we want to support all-or-nothing as well as partial updates? Or maybe reopen https://github.com/envoyproxy/envoy/issues/3079, I see even I was conflicted on this one.

Reply:
Opened #5739.